app/scripts/lib/metrics.js

```diff
@@ -186,6 +194,9 @@ define(function (require, exports, module) {
           }
```

Comment:
d45586ec4ca8eeece12ebc3332310ea024dc2fe5 made the change from clearing `events` and `timers` synchronously before the send to asynchronously, this is when we introduced `sendBeacon`. I wonder if there's any harm calling `clear` before `this._send` and not worrying about setting an `_isFlushing`.

What I worry about with the semaphorish approach is that we'll lose events - scenario is `sendBeacon` is not available so we fall back to an XHR POST request. While the XHR request is outstanding, events are added to the list. As soon as the XHR request completes, all events are cleared and the new ones are forgotten about.

Maybe like we talked about IRL if we get the filtered data, clear the events. If there's a failure sending, retry once, or at most twice. If events still aren't sent, oh well.